ml-agents/mlagents/trainers/poca/optimizer_torch.py

```diff
@@ -381,116 +382,110 @@ def _evaluate_by_sequence_team(
 
-            for groupmate_obs, team_action in zip(obs, actions):
+            for groupmate_obs, groupmate_action in zip(obs, actions):
                 seq_obs = []
-                for (_obs,) in groupmate_obs:
-                    first_seq_obs = _obs[start:end]
-                    seq_obs.append(first_seq_obs)
+                for _obs in groupmate_obs:
+                    sliced_seq_obs = _obs[start:end]
+                    seq_obs.append(sliced_seq_obs)
                 groupmate_seq_obs.append(seq_obs)
-                _act = team_action.slice(start, end)
+                _act = groupmate_action.slice(start, end)
                 groupmate_seq_act.append(_act)
 
             all_seq_obs = self_seq_obs + groupmate_seq_obs
             values, _value_mem = self.critic.critic_pass(
                 all_seq_obs, _value_mem, sequence_length=self.policy.sequence_length
             )
-            all_values = {
-                signal_name: [init_values[signal_name]] for signal_name in values.keys()
-            }
+            for signal_name, _val in values.items():
+                all_values[signal_name].append(_val)
 
             groupmate_obs_and_actions = (groupmate_seq_obs, groupmate_seq_act)
             baselines, _baseline_mem = self.critic.baseline(
                 self_seq_obs[0],
                 groupmate_obs_and_actions,
                 _baseline_mem,
-                sequence_length=first_seq_len,
+                sequence_length=self.policy.sequence_length,
+            )
+            for signal_name, _val in baselines.items():
+                all_baseline[signal_name].append(_val)
+
+        # Compute values for the potentially truncated initial sequence
+        last_seq_len = leftover
+        if last_seq_len > 0:
+            self_seq_obs = []
+            groupmate_seq_obs = []
+            groupmate_seq_act = []
+            seq_obs = []
+            for _self_obs in self_obs:
+                last_seq_obs = _self_obs[-last_seq_len:]
+                seq_obs.append(last_seq_obs)
+            self_seq_obs.append(seq_obs)
+
+            for groupmate_obs, groupmate_action in zip(obs, actions):
+                seq_obs = []
+                for _obs in groupmate_obs:
+                    last_seq_obs = _obs[-last_seq_len:]
+                    seq_obs.append(last_seq_obs)
+                groupmate_seq_obs.append(seq_obs)
+                _act = groupmate_action.slice(len(_obs) - last_seq_len, len(_obs))
+                groupmate_seq_act.append(_act)
+
+            # For the last sequence, the initial memory should be the one at the
+            # beginning of this trajectory.
+            seq_obs = []
+            last_seq_len = leftover
+            for _ in range(last_seq_len):
+                all_next_value_mem.append(ModelUtils.to_numpy(_value_mem.squeeze()))
+                all_next_baseline_mem.append(
+                    ModelUtils.to_numpy(_baseline_mem.squeeze())
```

Comment:
Why are we padding with this value and not NaN or zeros ?

Reply:
The sequence at this point isn't padded - it's just the leftover bit that would go into a new sequence. E.g. if there were three sequences of length "2" `[[s0, s1],[s2, s3],[s4]]`, at this point this is just `s4`. When we pad in the buffer, it will become [s4, 0], but not at this point. 

With that said, only the 1st memory of each sequence is used. I'm repeating the memory rather than using zeros or ones so we don't need to allocate a new numpy array every time.